### DISCUSSION TITLE
Favorite tools?

### DISCUSSION THREAD

**Discussion by sampsyo** [cucapra/latte21#47] [General]
Use this thread to paste links to your favorite language-based tools for accelerator design! Self-promotion is allowed, but also show off other stuff you use and love. ✨

**Comment by joaobispo** [comment]
Going to shamelessly plug a self-promotion here! I develop [Clava](https://github.com/specs-feup/clava/), an open-source C/C++ transpiler.

It doesn't actually fit as an accelerator design tool, but can be useful to analyze and transform C/C++/OpenCL code. And it has [CMake integration](https://github.com/specs-feup/clava/tree/master/CMake).

**Reply by joaobispo** [reply]
(more self promotion)

In one of the break rooms, Stephen Neuendorffer asked if we could apply the model that Clava uses (e.g. using JavaScript to analyze and transform ASTs), but instead of being over C/C++, doing that over an MLIR AST.

The answer is yes, we could! Clava is built on top of the LARA framework, which can be used to build this kind of transpilers. If the parser/AST is already available as a Java library, it is straightforward to create one of these transpilers (we have actually created several already, i.e. for Java, JavaScript, MATLAB). If the parser is not available as a Java library, it is not a problem (for instance, Clava uses Clang to parse C/C++ code), but it takes more effort.

**Comment by soronpo** [comment]
Self promoting the tool I only partially presented, DFiant HDL: https://dfianthdl.github.io/intro/first-look/
:wink:

**Comment by soronpo** [comment]
Really cool hardware description languages, 
Silice: https://github.com/sylefeb/Silice
PipelineC: https://github.com/JulianKemmerer/PipelineC

**Reply by JulianKemmerer** [reply]
Hey thanks for giving **PipelineC** a shout out! Loving to watch all the talks at LATTE this year!

Also - doubling down on **Silice** being cool

**Comment by angl-dev** [comment]
Another self-promotion here XD. Not a accelerator design tool, but an eFPGA generator/CAD flow: [PRGA](http://parallel.princeton.edu/prga/) ([GitHub](https://github.com/PrincetonUniversity/prga)). Could be an interesting target for all sorts of accelerator design tools :)

**Comment by angl-dev** [comment]
And my picks for accelerator/hardware design frameworks: [PyMTL3](https://github.com/pymtl/pymtl3), [HeteroCL](https://heterocl.csl.cornell.edu/).

**Reply by ekiwi** [reply]
Are there any cool example projects written in PyMTL that you know of? I know of the system level modelling capabilities and the class designed around PyMTL, but do people also use it for designs that they tape out?

**Reply by angl-dev** [reply]
I've collaborated with them on a recent tapeout. Cannot disclose the detail yet, but it works pretty well. Also [PyOCN](https://github.com/cornell-brg/pymtl3-net) may be worth checking out.

**Comment by stephenneuendorffer** [comment]
@benreynwar https://ptolemy.berkeley.edu/books/Systems/

**Comment by rachitnigam** [comment]
[Calyx](https://capra.cs.cornell.edu/calyx/) is a pass-based compiler infrastructure for compilers that generate hardware designs! We've got frontends for [Dahlia](capra.cs.cornell/dahlia) (an HLS-like language), a systolic array generator, NTT generator, and a WIP TVM frontend. Please let us know if you want to use it in your research!

**Reply by joaobispo** [reply]
This week we suggested to a MSc student we are advising to check Calyx and see if he could generate some code.

**Reply by ekiwi** [reply]
It is fairly easy to get running and the [documentation](https://capra.cs.cornell.edu/docs/calyx/) is pretty helpful. I recently set it up to show a demo during our group meeting.

**Reply by rachitnigam** [reply]
@joaobispo feel free to email us if you run into any problems! Thanks for demoing it @ekiwi!

**Reply by nmcp88** [reply]
Specifically on what @joaobispo said, we are trying to see if Calyx is a good IR, starting from instruction traces of native RISC-V code, transformed into CDFGs which we want to implement into hardware (e.g., tightly coupled loop accelerators etc). Seems like it might be straightforward to transform CDFGs which already represent arithmetic at a low level into the sub-languages of Calyx, no?
